search/search_repository.go

```diff
@@ -0,0 +1,151 @@
+
+	// This SQL query is used when search is performed across workitem ID ONLY.
+	testID = `select * from work_items WHERE to_tsvector('english', id::text || ' ') @@ to_tsquery('english',$1) and deleted_at is NULL`
+)
+
+// GormSearchRepository provides a Gorm based repository
+type GormSearchRepository struct {
+	ts  *models.GormTransactionSupport
+	wir *models.GormWorkItemTypeRepository
+}
+
+// NewGormSearchRepository creates a new search repository
+func NewGormSearchRepository(ts *models.GormTransactionSupport, wir *models.GormWorkItemTypeRepository) *GormSearchRepository {
+	return &GormSearchRepository{ts, wir}
+}
+
+func generateSearchQuery(q string) (string, error) {
+	return q, nil
+}
+
+func convertFromModel(wiType models.WorkItemType, workItem models.WorkItem) (*app.WorkItem, error) {
+	result := app.WorkItem{
+		ID:      strconv.FormatUint(workItem.ID, 10),
+		Type:    workItem.Type,
+		Version: workItem.Version,
+		Fields:  map[string]interface{}{}}
+
+	for name, field := range wiType.Fields {
+		var err error
+		result.Fields[name], err = field.ConvertFromModel(name, workItem.Fields[name])
+		if err != nil {
+			return nil, err
+		}
+	}
+
+	return &result, nil
+}
+
+func (r *GormSearchRepository) loadTypeFromDB(ctx context.Context, name string) (*models.WorkItemType, error) {
+	log.Printf("loading work item type %s", name)
+	res := models.WorkItemType{}
+
+	if r.ts.TX().Where("name=?", name).First(&res).RecordNotFound() {
+		log.Printf("not found, res=%v", res)
+		return nil, NotFoundError{"work item type", name}
+	}
+	if err := r.ts.TX().Error; err != nil {
+		return nil, InternalError{simpleError{err.Error()}}
+	}
+
+	return &res, nil
+}
+
+//searchKeyword defines how a decomposed raw search query will look like
+type searchKeyword struct {
+	id    []string
+	words []string
+}
+
+// parseSearchString accepts a raw string and generates a searchKeyword object
+func parseSearchString(rawSearchString string) searchKeyword {
+	// TODO remove special characters and exclaimations if any
+	rawSearchString = strings.ToLower(rawSearchString)
+	parts := strings.Split(rawSearchString, " ")
+	var res searchKeyword
+	for _, part := range parts {
+		if strings.HasPrefix(part, "id:") {
+			res.id = append(res.id, strings.Trim(part, "id:"))
+		} else {
+			res.words = append(res.words, part)
+		}
+	}
+	return res
+}
+
+func generateSQLSearchInfo(keywords searchKeyword) (sqlQuery string, sqlParameter string) {
+	idStr := strings.Join(keywords.id, " & ")
+	wordStr := strings.Join(keywords.words, " & ")
+	searchQuery := testText
+
+	if len(keywords.id) == 1 && len(keywords.words) == 0 {
+		// If the search string is of the form "id:2647326482" then we perform
+		// search only on the ID, else we do a full text search.
+		// Is "id:45453 id:43234" be valid ? NO, because the no row can have 2 IDs.
+		searchQuery = testID
+	}
+	return searchQuery, idStr + wordStr
```

Comment:
@tsmaeder  thanks for pointing out. ! I missed that.
@sbose78 checked, that was needed thanks.